pyomo/contrib/doe/measurements.py

```diff
@@ -25,252 +25,270 @@
-        flatten_measure_name: flattened measurement names. For e.g., flattenning {'C':{'CA': 10, 'CB': 1, 'CC': 2}} will be 'C_index_CA', ..., 'C_index_CC'.
-        variance:
-            a ``dict``, keys are measurement variable names, values are a dictionary, keys are its extra index name,
-            values are its variance as a scalar number.
-            For e.g., for the kinetics illustrative example, it should be {'C':{'CA': 10, 'CB': 1, 'CC': 2}}.
-            If given None, the default is {'C':{'CA': 1, 'CB': 1, 'CC': 1}}.
-            If there is no extra index, it is a dict, keys are measurement variable names, values are its variance as a scalar number.
-        name_and_index:
-            a dictionary, keys are measurement names, values are a list of extra indexes.
+
+        for i, n in enumerate(var_name):
+            name_data = str(n)
+
+            # first combine all indexes into a list 
+            all_index_list = [] # contains all index lists
+            if extra_index:
+                for index_list in extra_index[i]: 
+                    all_index_list.append(index_list)
+            if time_index:
+                all_index_list.append(time_index[i])
+
+            # all idnex list for one variable, such as ["CA", 10, 1]
+            all_index_for_var = list(itertools.product(*all_index_list))
+
+            for lst in all_index_for_var:
+                name1 = name_data+"["
+                for i, idx in enumerate(lst):
+                    name1 += str(idx)
+
+                    # if i is the last index, close the []. if not, add a "," for the next index. 
+                    if i==len(lst)-1:
+                        name1 += "]"
+                    else:
+                        name1 += ","
+
+                self.special_set.append(name1)
+
+
+    def _names(self, var_name, extra_index, time_index, values, upper_bound, lower_bound, check_values_bounds=False):
+        """
+        Check if the measurement information provided are valid to use. 
+        """
+        num_var = len(var_name)
+        
+        if extra_index and len(extra_index) != num_var: 
+            raise ValueError("Extra_index is of different length with var_name. ")
+
+        if len(time_index) != num_var:
+            raise ValueError("time_index is of different length with var_name.")
+
+        if check_values_bounds:
+            num_name = len(self.special_set)
+
+            if values and len(values) != num_name:
+                raise ValueError("Values is of different length with var_name.")
+            
+            if upper_bound and type(upper_bound)==list and len(upper_bound)!= num_name:
+                raise ValueError("Upperbounds is of different length with var_name.")
+            
+            if lower_bound and type(lower_bound)==list and len(lower_bound)!= num_name:
+                raise ValueError("Lowerbounds is of different length with var_name.")
+
+    
+class Measurements(SpecialSet):
```

Comment:
I think we should extend this class to also allow for `uncertainty` to be stored in addition to `value`. This will give us a convenient way to assume a diagonal covariance structure but specify unique uncertainties for each measurement.